### PR TITLE
[dbnode] Wide query iterator implementation

### PR DESCRIPTION
**Pull request by arnikola**
Provides an iterator over wide query results; resubmission off #2870

### CODE REVIEW

File: src/dbnode/storage/wide/types.go [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+// QueryIterator is a wide query iterator.
+type QueryIterator interface {
+	wideIter
+
+	// Shards are the shards this iterator is responsible for.
+	Shards() []uint32
+	// Current is the current shard iterator, invalidated on calls to Next().
+	Current() QueryShardIterator
+	// ShardIter is used to write into the given shard.
+	ShardIter(shard uint32) (QueryShardIterator, error)
```
Comment (linasm) [inline src/dbnode/storage/wide/types.go:83]: This method name does not communicate enough about what it does. Guess adding some verb would help.

File: src/dbnode/storage/wide/wide_query_iter.go [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+	i.state.Lock()
+	err := i.state.err
+	i.state.Unlock()
+	return err
+}
+
+func (i *wideQueryIterator) Close() {
+}
+
+const shardIterRecordsBuffer = 1024
```
Comment (linasm) [inline src/dbnode/storage/wide/wide_query_iter.go:161]:
```suggestion
const shardIterRecordsBufferSize = 1024
```

File: src/dbnode/storage/wide/wide_query_iter.go [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+
+func (i *wideQueryShardIterator) BlockStart() time.Time {
+	return i.blockStart
+}
+
+func (i *wideQueryShardIterator) setDone() {
+	close(i.records)
+}
+
+// nolint: hugeParam
```
Comment (linasm) [inline src/dbnode/storage/wide/wide_query_iter.go:219]: Is this `nolint` needed?

File: src/dbnode/storage/wide/wide_query_iter.go [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+	var (
+		qr  wideQueryShardIteratorQueuedRecord
+		err error
+	)
+
+	// TODO: transactionally copy the ID + tags + anything else in one go
+	// otherwise the fixed buffer manager might run out of mem and wait
+	// for another buffer to be available but the existing buffer cannot
+	// be released since one field here has taken a copy and needs to be
+	// returned for entire underlying buffer to be released.
```
Comment (linasm) [inline src/dbnode/storage/wide/wide_query_iter.go:230]: When do we plan to implement this?

File: src/dbnode/storage/wide/wide_query_iter.go [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+func NewQueryIterator(
+	blockStart time.Time,
+	shards []uint32,
+	opts Options,
+) QueryIterator {
+	return &wideQueryIterator{
+		fixedBufferMgr: newFixedBufferManager(opts),
+		iters:          make(chan *wideQueryShardIterator, len(shards)),
+		writeShard:     shardNotSet,
+		opts:           opts,
```
Comment (linasm) [inline src/dbnode/storage/wide/wide_query_iter.go:76]: `blockStart` is not being set.

File: src/dbnode/storage/wide/wide_query_iter.go [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+
+func (i *wideQueryIterator) Err() error {
+	i.state.Lock()
+	err := i.state.err
+	i.state.Unlock()
+	return err
+}
+
+func (i *wideQueryIterator) Close() {
+}
```
Comment (linasm) [inline src/dbnode/storage/wide/wide_query_iter.go:159]: Wouldn't we want to close the `i.iters` channel?

File: src/dbnode/storage/wide/wide_query_iter.go [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+
+func (i *wideQueryShardIterator) Close() {
+}
+
+var _ QuerySeriesIterator = (*wideQuerySeriesIterator)(nil)
+
+type wideQuerySeriesIterator struct {
+	record      wideQueryShardIteratorQueuedRecord
+	reuseableID *ident.ReuseableBytesID
```
Comment (linasm) [inline src/dbnode/storage/wide/wide_query_iter.go:288]: Typos:
```suggestion
	reusableID *ident.ReusableBytesID
```

File: src/dbnode/storage/wide/wide_query_iter_prop_test.go [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+)
+
+type expectedRecord struct {
+	vals         []float64
+	expectedTags []byte
+	expectedID   []byte
+}
+
+type testRecord struct {
+	record   ShardIteratorRecord
```
Comment (linasm) [inline src/dbnode/storage/wide/wide_query_iter_prop_test.go:54]: Nit: maybe
```suggestion
	actual   ShardIteratorRecord
```

File: src/dbnode/storage/wide/types.go [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+	Close()
+}
+
+// CrossBlockSeries is a cross-block series.
+type CrossBlockSeries struct {
+	ID          ident.ID
+	EncodedTags ts.EncodedTags
+	Datapoints  []ts.Datapoint
+	Unit        xtime.Unit
+	Annotation  ts.Annotation
```
[the remark this answers is not in the document]
Comment (linasm) [reply src/dbnode/storage/wide/types.go:146]: Also looks like these fields are not in any use.

File: src/dbnode/storage/wide/wide_query_iter_prop_test.go [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+		}
+
+		return testRecord{
+			record: ShardIteratorRecord{
+				EncodedTags: []byte(tags),
+				ID:          []byte(id),
+				Data:        encodedData,
+			},
+			expected: expectedRecord{
+				vals:         append(make([]float64, 0, len(vals)), vals...),
```
Comment (linasm) [inline src/dbnode/storage/wide/wide_query_iter_prop_test.go:104]: Nit: don't think we need to take a copy here.

File: src/dbnode/storage/wide/wide_query_iter_prop_test.go [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+			seg.Head.IncRef()
+			encodedData = append(encodedData, seg.Head.Bytes()...)
+			seg.Head.DecRef()
+		}
+
+		if seg.Tail != nil {
+			seg.Tail.IncRef()
+			encodedData = append(encodedData, seg.Tail.Bytes()...)
+			seg.Tail.DecRef()
+		}
```
Comment (linasm) [inline src/dbnode/storage/wide/wide_query_iter_prop_test.go:95]: Nit: might just skip those Inc/DecRef in test code?
Probably needs `seg.Finalize()` if you want to get the full life cycle.

File: src/dbnode/storage/wide/wide_query_iter_prop_test.go [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+}
+
+func TestWideQueryIter(t *testing.T) {
+	var (
+		parameters = gopter.DefaultTestParameters()
+		seed       = time.Now().UnixNano()
+		props      = gopter.NewProperties(parameters)
+		reporter   = gopter.NewFormatedReporter(true, 80, os.Stdout)
+
+		// opts         = NewOptions().SetDecodingOptions(decodingOpts)
```
Comment (linasm) [inline src/dbnode/storage/wide/wide_query_iter_prop_test.go:126]: Drop?

File: src/dbnode/storage/wide/wide_query_iter_prop_test.go (hunk 8 above)
Comment (linasm) [reply src/dbnode/storage/wide/wide_query_iter_prop_test.go:54]: OTOH, `record` appears to be unused.

File: src/dbnode/storage/wide/wide_query_iter_prop_test.go [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+
+				iter.Close()
+
+				if len(shardVals) != len(genRecords) {
+					return false, fmt.Errorf("expected %d vals, got %d",
+						len(shardVals), len(genRecords))
+				}
+
+				for idx, record := range genRecords {
+					if !assert.ObjectsAreEqual(shardVals[idx], record.expected.vals) {
```
Comment (linasm) [inline src/dbnode/storage/wide/wide_query_iter_prop_test.go:205]: It takes expected as the first argument.

File: src/dbnode/storage/wide/wide_query_iter_test.go [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+	iter.reset(wideQueryShardIteratorQueuedRecord{
+		id:                []byte("id"),
+		borrowID:          singleFinalizeBorrower(),
+		encodedTags:       []byte("tags"),
+		borrowEncodedTags: singleFinalizeBorrower(),
+		data:              []byte("data"),
+		borrowData:        singleFinalizeBorrower(),
+	})
+
+	assert.Equal(t, iter.SeriesMetadata(), SeriesMetadata{
```
Comment (linasm) [inline src/dbnode/storage/wide/wide_query_iter_test.go:81]: Swap expected vs actual.

File: src/dbnode/storage/wide/wide_query_iter_test.go [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+		ID:          ident.BytesID([]byte("id")),
+		EncodedTags: ts.EncodedTags([]byte("tags")),
```
Comment (linasm) [inline src/dbnode/storage/wide/wide_query_iter_test.go:83]: Nit:
```suggestion
		ID:          []byte("id"),
		EncodedTags: []byte("tags"),
```

**Review by linasm — CHANGES_REQUESTED** [review]
My biggest concern - are we so desperate that we need to implement another memory manager?

File: src/dbnode/storage/wide/fixed_buffer.go [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+
+	// All copies mutate.
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	var ok bool
+	if m.curr == nil {
+		m.curr, ok = m.pool.getWithTimeout(m.timeout)
+		if !ok {
+			return nil, fixedBufferBorrowNoopGlobal, errTimeout
```
[the remark this answers is not in the document]
Comment (arnikola) [reply src/dbnode/storage/wide/fixed_buffer.go:67]: We want this to be strictly bounded more than anything; may add a strict mode that could toggle this behaviour

**Pull request closed without merging**